component/src/main/java/io/siddhi/extension/io/grpc/source/GrpcServiceSource.java

```diff
@@ -148,14 +150,23 @@ public void process(Event request,
         }
     }
 
-    class ServiceSourceTimeoutChecker implements Runnable {
+    class ServiceSourceTimeoutChecker extends TimerTask {
+        private String messageId;
+        private long receivedTime;
+
+        public ServiceSourceTimeoutChecker(String messageId, long receivedTime) {
+            this.messageId = messageId;
+            this.receivedTime = receivedTime;
+        }
+
         @Override
         public void run() {
-            for (String messageId: timestampOfMessageIds.keySet()) {
-                if (timestampOfMessageIds.get(messageId) < siddhiAppContext.getTimestampGenerator().currentTime() -
-                        serviceTimeout) {
-                    streamObserverMap.get(messageId).onError(new io.grpc.StatusRuntimeException(
+            if (receivedTime < siddhiAppContext.getTimestampGenerator().currentTime() - serviceTimeout) {
```

Comment:
We need to wait in sleep while loop till this condition is met (if the timer has released the value earlier)

Reply:
I am starting this thread with an initial delay of service.timeout. So anyway this condition will be checked after service.timeout

Reply:
So why we need to do the if check?  
siddhiAppContext.getTimestampGenerator().currentTime() can be varied based on the playback so we can always assume that. Implement a while loop to handle this.

Reply:
we don't need the if check. After the service timeout the existence of that stream observer will be enough to decide to send error. Will remove if